M2/Macaulay2/m2/html.m2

```diff
@@ -162,40 +164,59 @@ html TO2  := x -> (
 ----------------------------------------------------------------------------
 
-html Thing := htmlLiteral @@ (x -> "$" | texMath x | "$") -- by default, we use math mode tex (as opposed to actual html)
+texHtml =
+html Thing := x -> "$" | htmlLiteral texMath x | "$" -- by default, we use math mode tex (as opposed to actual html)
 html Nothing := x -> ""
 
 -- text stuff: we use html instead of tex, much faster (and better spacing)
+-- TODO: rewrite next 2 using hypertext, just like the ones after, once PRE spacing/line break issues are fixed
 html Net := n -> concatenate("<pre style=\"display:inline-table;text-align:left;vertical-align:",
     toString(if #n>0 then 100*(height n-1) else 0), "%\">\n", -- the % is relative to line-height
     apply(unstack n, x-> htmlLiteral x | "<br/>"), "</pre>")
 html String := x -> concatenate("<pre style=\"display:inline\">\n", htmlLiteral x,
     if #x>0 and last x === "\n" then "\n" else "", -- fix for html ignoring trailing \n
     "</pre>")
-html Descent := x -> concatenate("<span style=\"display:inline-table;text-align:left\">\n", apply(sortByName pairs x,
+-- a few types are just strings
+hypertext Descent := x -> SPAN prepend( "style" => "display:inline-table;text-align:left", -- TODO move style to CSS
+    deepSplice apply(sortByName pairs x,
      (k,v) -> (
 	  if #v === 0
-	  then html k
-	  else html k | " : " | html v
-	  ) | "<br/>"), "</span>")
-html Time := x -> html x#1 | html DIV ("-- ", toString x#0, " seconds")
--- a few types are just strings
-html Command :=
-html File :=
-html IndeterminateNumber :=
-html Boolean :=
-html Function :=
-html FilePosition :=
-html Manipulator :=
-html Dictionary := html @@ toString
-html Type := x -> if x.?texMath then "$"|x.texMath|"$" else html toString x
+	  then k
+	  else (k, " : ", v)
+	  , BR{})))
+hypertext Time := x -> DIV { x#1, DIV ("-- ", toString x#0, " seconds", "class" => "token comment") }
+TTc = c -> x -> TT {toString x,"class"=>"token "|c}
+hypertext Pseudocode :=
+hypertext CompiledFunctionBody := TTc "function"
+hypertext Command :=
+hypertext FunctionBody :=
+hypertext Function := f -> TT deepSplice {
+    if hasAttribute(f,ReverseDictionary) then toString getAttribute(f,ReverseDictionary) else (
+	t := locate if instance(f,Command) then f#0 else f;
+	"-*",
+	SPAN class f,
+	if t =!= null then ("[", SPAN t, "]"),
+	"*-"
+	),
+    "class"=>"token function"
+    }
+hypertext Package :=
+hypertext File :=
+hypertext IndeterminateNumber :=
+hypertext Manipulator :=
+hypertext Boolean := TTc "constant"
+hypertext Type :=
+hypertext FilePosition :=
+hypertext Dictionary := TTc "class-name"
+--hypertext VerticalList         := x -> UL apply(x, y -> new LI from hold y)
+--hypertext NumberedVerticalList := x -> OL apply(x, y -> new LI from hold y)
+scan(methods hypertext, (h,t) -> html t := html @@ hypertext);
 -- except not these descendants
 html Monoid :=
 html RingFamily :=
-html Ring := lookup(html,Thing)
+html Ring := texHtml
```

Comment:
Maybe it's just me, but `lookup(html, Thing)` is more informative to me than `texHtml`.

Alternatively, why not change it into this?
```m2
html Monoid :=
html RingFamily :=
html Ring :=
html Thing := x -> if lookup(hypertext, class x) then html hypertext x else "$" | htmlLiteral texMath x | "$"
```

Reply:
sure

Reply:
(This would eliminate the need for the `scan` a few lines higher also.)

Reply:
wait no that doesn't work.  for say a `Ring` the lookup will be nonnull and yet we do want to `texMath` it, not go through `hypertext`.

Reply:
What does `hypertext Ring` do? Maybe that should also behave differently then?

Reply:
well, it's inherited from `hypertext Type`...